### PR TITLE
Task 4

### PR DESCRIPTION
**Pull request by DzmitryU**
**What was done?**
1. Created RDS PostgreSQL DB
2. Products are fetched by Lambda from RDS
3. Added endpoint for products creation
4. Front end is integrated with Lambda and is getting products from RDS: https://dp11r8ce0ltcp.cloudfront.net/

**Additional:**
1.  Used Joi for product input validation during creation (return 400 in case of invalid product)
2. 500 Error is returned in case of any type of errors (handling on data layers throws an error up after logging)
3. Added console log to all lambdas handlers input
4. Implemented transaction logic for product creation

**Url:** https://64rj2xj4v3.execute-api.eu-west-1.amazonaws.com/dev/products/
**FE pr:** https://github.com/DzmitryU/nodejs-aws-fe/pull/3

### CODE REVIEW

File: product-service/src/dataStore/pg-client.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        client.end();
+    }
+}
+
+export const get = async (id): Promise<Product> => {
+    const client = new Client(dbOptions);
+    await client.connect();
+
+    try {
+        const script = `select ${productFields} from products, stocks where id = '${id}' and id = product_id`;
```
Comment (LeanidShutau) [inline product-service/src/dataStore/pg-client.ts:46]: Risk of SQL injection. Even if you validate `id` with Joi, this function is independent from it, so by its own behavior this query is vulnerable.

File: product-service/src/dataStore/pg-client.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+
+export const insert = async (product: Product): Promise<Product> => {
+    const client = new Client(dbOptions);
+    await client.connect();
+
+    try {
+        await client.query('BEGIN');
+        const productsScript =
+            `insert into products (description, price, title) ` +
+            `values ('${product.description}',${product.price},'${product.title}')` +
```
Comment (LeanidShutau) [inline product-service/src/dataStore/pg-client.ts:68]: Same SQL injection here but these arguments are not sanitized so this query is vulnerable right now.

**Review by LeanidShutau — APPROVED** [review]
All main and additional tasks are done.\
It would be nice if you fix SQL injections but it's not required by the task so the total score is 9/9.

File: product-service/.gitignore [hunk 3]
```diff
@@ -6,4 +6,5 @@ jspm_packages
 .serverless
 
 # Webpack directories
-.webpack
\ No newline at end of file
+.webpack
+/.env
```
Comment (kotmatpockuh) [inline product-service/.gitignore]: / - this is redundant ;)

File: product-service/serverless.ts [hunk 4]
```diff
@@ -24,6 +24,11 @@ const serverlessConfiguration: Serverless = {
         },
         environment: {
             AWS_NODEJS_CONNECTION_REUSE_ENABLED: '1',
+            PG_HOST: '${env:PG_HOST}',
+            PG_PORT: '${env:PG_PORT}',
+            PG_DATABASE: '${env:PG_DATABASE}',
```
Comment (kotmatpockuh) [inline product-service/serverless.ts]: `env:PG_DATABASE` - you can use it directly in the place, see no reason to have it also in serverless.ts 🤔

**Review by kotmatpockuh — APPROVED** [review]

**Comment by kotmatpockuh** [comment]
9/9

ps: `Added console log to all lambdas handlers input` (looks like we need to output also request body etc ;))

File: product-service/src/dataStore/pg-client.ts (hunk 1 above)
Comment (DzmitryU) [reply product-service/src/dataStore/pg-client.ts:46]: What's the best security approach? To check all insert params directly before injection into query?

File: product-service/serverless.ts (hunk 4 above)
Comment (DzmitryU) [reply product-service/serverless.ts]: I believe that describing env vars in serverless.ts would configure values used in Lambda after deployment, wouldn't it?

Comment (kotmatpockuh) [reply product-service/serverless.ts]: Looks like yes & no. You can easily use .env vars just in the handler and all magic will be resolved...

**Pull request merged**